src/HOCON/HoconParser.cs

```diff
@@ -0,0 +1,1011 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="HoconParser.cs" company="Hocon Project">
+//     Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
+//     Copyright (C) 2013-2015 Akka.NET project <https://github.com/akkadotnet/Hocon>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Hocon
+{
+    public delegate string HoconIncludeCallback(HoconCallbackType callbackType, string value);
+    public delegate Task<string> HoconIncludeCallbackAsync(HoconCallbackType callbackType, string value);
```

Comment:
The main purpose is to have it non-blocking for "include url()" because it potentially has to wait a long time and large Hocon config file cases because the current Tokenizer implementation does not support string Stream. Since I'm trying to make Hocon as standalone and as generic as possible, I thought making everything async is the best move.

Reply:
My big concern here is that how much this feature justifies all of this? Tbh. I never knew that you can include URL or file in you HOCON files and I'm using it for the last 4 years. On the other side parsing and text processing can be made very lightweight if we won't try to make it all async. My gut feeling is that 2 heaviest parts of `Parse...Async` methods are throwing exceptions and actually calling await/returning a Task.

From personal experience: few weeks ago I've wrote a CSV parser using the latests bits of .NET low-level API. After benchmarking it on 1.5GB files it turned out that over 80% of memory and CPU was used for async state machine related to async loading of files and 8% was used by garbage collector to recycle those tasks (the rest of parsing pipeline was not allocating anything).

Reply:
Well, I can always remove all the async await, its pretty easy to do since none of my code are actually asynchronous. It was there for completeness, since I wanted to support the `required url()` spec. 
The whole point of this exercise is trying to support the full spec and make Hocon a real standalone configuration library outside of Akka (ie. just like NewtonSoft Json), so I'm just covering all the bases.

Reply:
Parsing HOCON should be a boot up process, it seems farfetched that there would be so many calls to the the parser that it justifies a complete async API IMO.

If the purpose is to make it more maintainable, aim to make it smaller/less code instead of adding things with low end user value.

my 2 cents

Reply:
Refactored it so that only the include callback is async, would this be a better compromise?